07.md

```diff
@@ -20,6 +20,8 @@ Aside from these two basic above, the following functions can also be implemente
 async window.nostr.getRelays(): { [url: string]: {read: boolean, write: boolean} } // returns a basic map of relay urls to relay policies
 async window.nostr.nip04.encrypt(pubkey, plaintext): string // returns ciphertext and iv as specified in nip-04 (deprecated)
 async window.nostr.nip04.decrypt(pubkey, ciphertext): string // takes ciphertext and iv as specified in nip-04 (deprecated)
+async window.nostr.nip44.encrypt([[pubkey1, plaintext1], [pubkey2, plaintext2], ...]): string[] // takes array of [pubkey, plaintext] tuples, returns array of ciphertexts as specified in nip-44
+async window.nostr.nip44.decrypt([[pubkey1, ciphertext1], [pubkey2, ciphertext2], ...]): string[] // takes array of [pubkey, ciphertext] tuples, returns array of plaintexts as specified in nip-44
```

Comment:
I will also just add real quick, that letting you encrypt/decrypt multiple at once leads to the possibility that the signer will return an array with a different length than the one you passed in. It's not really a big deal since clients need to validate the response anyway.

Reply:
if theres an error, I would think that returned array should retain the same number of values and just put a blank value for the one that err'ed. would probably be good to explicitly say that here.

do you think thats a valid way of handling it?

Reply:
If any individual message has an error, the whole function call should throw.

Reply:
The more I think about it, the more I think it's actually a problem. The client has no way to validate messages before sending them to the signer, right? So any invalid text can cause the signer to throw (or return a rejected promise).

First I would question if we really need to support decrypting multiple messages. Is it just because Alby prompts every time?

If we really need to do it, it should have a return type of `Promise<string>[]` instead of `Promise<string[]>` so individual calls can be either resolved or rejected.

Reply:
no, it is because for the NIP44 calls in NIP46 it would be extremely inefficient to fetch one decryption at a time. so it is better for NIP46 to have these calls be arrays, and we want the NIP07 calls to be consistent with the NIP46 calls. hence why the NIP44 calls have array parameters.

can you return a promise from a browser extension? returning a promise definitely wouldn't work for NIP46 either, which we are trying to stay consistent with.

Reply:
maybe the array returned should house objects like:
```json
{
  "result": "result string"
  "error": "error description"
}
```

don't love wrapping everything in a object though. maybe only the item that err'ed should return an object (rather than a string)? that should be pretty parseable right?

Reply:
I would say consistency with NIP-46 is not a good enough reason for this to be an array. It's okay for them to be different if there's a good reason. This is a JavaScript API. No other JavaScript API has you pass in an array of multiple arguments. It's an anti-pattern. We should change it to do only one at a time. NIP-46 can do it the complicated way. Let's not turn something simple into something complex just because another thing is complex.

Reply:
do you have anything I can read on JavaScript API patterns?

anti-patterns with other nostr functionality (ie NIP46) was my main concern here and imo is the most important pattern to adhere to. (I don't think I'm the only one who feels this way, see https://github.com/nostr-protocol/nips/pull/940#issuecomment-1952150248, and https://github.com/nostr-protocol/nips/pull/940#issuecomment-1952770826)

besides, I do think the array argument is beneficial to NIP07 as well as it allows browser extension signers to group permission requests if trying to create a better request pop up flow for the user. otherwise browsers are forced to prompt the user for each request individually (in the scenario that the user doesn't want to grant blanket decryption/encryption approval to the browser signer).

Reply:
Check my NIP-07 TypeScript interface: https://gitlab.com/soapbox-pub/NSpec/-/blob/main/interfaces/NostrSigner.ts

And here is an implementation, NSecSigner: https://gitlab.com/soapbox-pub/NSpec/-/blob/main/src/NSecSigner.ts

Usage:

```ts
const signer = new NSecSigner(secretKey); // We pass the secretKey when constructing the object

// Now it's API-compatible with NIP-07 and can even polyfill `window.nostr`
await signer.getPublicKey();
await signer.signEvent(event);
```

As another example, `NSeedSigner` lets you pass a mnemonic seed phrase, derives the secretKey, and then uses `NSecSigner` internally: https://gitlab.com/soapbox-pub/NSpec/-/blob/main/src/NSeedSigner.ts

Finally, `SoapboxSigner` is an application-specific signer that chooses a signer based on the user's preferences: https://gitlab.com/soapbox-pub/soapbox/-/blob/ditto-auth/src/features/nostr/SoapboxSigner.ts

Believe me, I signer. I'm very invested in the signers. I get it. And so the question is, can you implement NIP-46 using the `NostrSigner` interface?

The answer is yes, even if NIP-46 uses arrays and NIP-07 uses atomic calls. You would simply throttle the NIP-07 function in the Signer implementation, and then batch them into a single NIP-46 call. NIP-46 would then return an array of results object like you mentioned, and the Signer implementation would resolve or reject the promises.

Here is a very rough pseudocode:

```ts
export class NConnectSigner extends EventTarget implements NostrSigner {
  #nip46: NIP46;
  #queue: string[][];

  constructor(pubkey: string, relays?: string[], delay?: number) {
    this.#nip46 = new NIP46(this.#pubkey, this.#relays);

    setInterval(() => {
      this.#batchAndSendCalls();
    }, delay ?? 100);

    super();
  }

  async getPublicKey(): Promise<string> {
    return this.#nip46.getPublicKey();
  }

  async signEvent(event: Omit<NostrEvent, 'id' | 'pubkey' | 'sig'>): Promise<NostrEvent> {
    return this.#nip46.signEvent(event);
  }

  readonly nip44 = {
    encrypt: async (pubkey: string, plaintext: string): Promise<string> => {
      this.#queue.push(['encrypt', pubkey, plaintext]);

      return new Promise((resolve) => {
        this.addEventListener('encrypt', resolve);
      })
    },

    decrypt: async (pubkey: string, ciphertext: string): Promise<string> => {
      this.#queue.push(['decrypt', pubkey, ciphertext]);

      return new Promise((resolve) => {
        this.addEventListener('decrypt', resolve);
      })
    },
  };

  #batchAndSendCalls() {
    for (const [method, ...args] of this.#queue) {
      this.#nip46[method](...args).then((result) => {
        this.dispatchEvent(new CustomEvent(method, { detail: result }));
      });
    }
  }
}
```

The point is, NIP-46 is a different beast than NIP-07. NIP-07 defines a JavaScript API. We should focus on creating the ideal JavaScript API. The way you future-proof an API is by making it as simple as possible.

Reply:
Btw, what I am suggesting here is something that could open the door to nip46 batching multiple types of different calls. So you could sign an event and decrypt stuff in the same call. This is why the JS API should be atomic.

Reply:
I still think its more important to create the ideal nostr API than javascript API. why should we force throttling, batching and creating a bunch of promises just so nip07 signer and nip46 signer can both use the same interface? if theres a real tangible reason I'd be more amenable, but just to follow some arbitrary javascript api standards (which I would still like to read more about if you have a resource) seems like not a good enough reason to increase implementation complexity.

>The way to future-proof an API is by making it as simple as possible, NOT to make it flexible for existing systems.

we are hardly making this flexible, just allowing multiple calls but for very straightforward and single use case. plus this arbitrary desire for atomic simplicity directly leads to real, known complexity on implementation of signer integration. 

>Btw, what I am suggesting here is something that could open the door to nip46 batching multiple types of different calls. So you could sign an event and decrypt stuff in the same call. This is why the JS API should be atomic.

having atomic calls, rather than arrays, is not a prerequisite for nip46 batching calls. plus the most useful and wanted type of batching is batching encryption and decryption requests. so why implement an api in its least useful form so that maybe someday we can implement something on top that makes it more useful? 

also, it sounds like you are also advocating for atomic nip46 as well? I understand the method call structure may not be standard, but imo making the method call practical and enabling good user experience is more important.

Reply:
I think I mostly agree with Alex here that NIP 46 should include batching, but software interfaces should be simple with throttle/batch behind the scenes. It would make sense in theory for the NIP 07 interface to match the simple js interface, and have the extension do batching, but from looking at Alby's code I think that might require a lot more involved re-write.

Implementation aside, which is not a good argument, is there ever a time when a NIP 07 should do batching? Is there ever a time when NIP 07 would not be talking to a browser extension? Probably not.

That said, one thing we haven't talked much about is getting rid of NIP 07 entirely, and making it just an implementation of NIP 46. 46 currently relies on relays explicitly, but if you could separate the interface/api and architecture, you could provide a NIP 46 signer in an extension, over http, etc. That would sort of make this a moot point since NIP 07 would no longer exist.

Reply:
What does Alby do about failed decrypt calls at the moment, when the failed element is in the middle of the array? Does it return `undefined` or return a rejected promise for the whole function? Because I think rejecting the whole promise is the correct idiomatic JavaScript way, but it would make the API totally useless if they did that.

Reply:
> from looking at Alby's code I think that might require a lot more involved re-write.

That doesn't make sense. Alby doesn't have to do batching unless it's converting calls to NIP-46. It would be less code on their part to only decrypt the one. They can even update their code with a one-liner: `nip44.decrypt = (...args) => _oldDecrypt([...args])`

Reply:
Found the Alby MR and commented here: https://github.com/getAlby/lightning-browser-extension/pull/2653#pullrequestreview-1891078523

Reply:
> What does Alby do about failed decrypt calls at the moment

That's a super good point, it throws.

> Alby doesn't have to do batching unless it's converting calls to NIP-46

Alby may want to do batching for UX rather than performance reasons. So instead of 300 pop ups, they should really show "Coracle wants to decrypt 300 messages. ok?".

Reply:
🤔 I might agree with you. I know @fiatjaf currently wants the interface the same. Maybe he would be convinced by your argument.